File: packages/dashboard/server/routers/reports.ts
// ReportEnvironmentalHazard boundary (SRS CP-C3)
// public users submit hazard reports, admins triage, operators see stats

import { hazardReports, auditLogs, alertSubscriptions } from '@scemas/db/schema'
import { CreateHazardReportSchema, UpdateHazardReportStatusSchema } from '@scemas/types'
import { and, count, desc, eq, inArray } from 'drizzle-orm'
import { z } from 'zod'
import { normalizeZoneId } from '@/lib/zones'
import { router, publicProcedure, protectedProcedure, adminProcedure } from '../trpc'

export const reportsRouter = router({
  submit: publicProcedure.input(CreateHazardReportSchema).mutation(async ({ input, ctx }) => {
    const [report] = await ctx.db
      .insert(hazardReports)
      .values({
        zone: normalizeZoneId(input.zone) ?? input.zone,
        category: input.category,
        description: input.description,
        contactEmail: input.contactEmail ?? null,
        reportedBy: ctx.user?.id ?? null,
      })
      .returning({ id: hazardReports.id, status: hazardReports.status })

    await ctx.db.insert(auditLogs).values({
      userId: ctx.user?.id ?? null,
      action: 'report.submitted',
      details: { reportId: report.id, zone: input.zone, category: input.category },
    })

    return report
  }),

  list: protectedProcedure
    .input(
      z.object({
        status: z.string().optional(),
        zone: z.string().optional(),
        limit: z.number().int().min(1).max(200).default(50),
        offset: z.number().int().min(0).default(0),
      }),
    )
    .query(async ({ input, ctx }) => {
      const conditions = []

      if (input.status) {
        conditions.push(eq(hazardReports.status, input.status))
      }

      if (input.zone) {
        conditions.push(eq(hazardReports.zone, input.zone))
      }

      // operators only see reports for their subscribed zones
      if (ctx.user.role !== 'admin') {
        const subscription = await ctx.db.query.alertSubscriptions.findFirst({
          where: eq(alertSubscriptions.userId, ctx.user.id),
        })
        const subscribedZones = subscription?.zones ?? []
        if (subscribedZones.length > 0) {
          conditions.push(inArray(hazardReports.zone, subscribedZones))
        }
      }

      const where = conditions.length > 0 ? and(...conditions) : undefined

      return ctx.db.query.hazardReports.findMany({
        where,
        orderBy: [desc(hazardReports.createdAt)],
        limit: input.limit,
        offset: input.offset,
      })
    }),

  get: protectedProcedure
    .input(z.object({ id: z.string().uuid() }))
    .query(async ({ input, ctx }) => {
      return ctx.db.query.hazardReports.findFirst({
        where: eq(hazardReports.id, input.id),
      })
    }),

  updateStatus: adminProcedure
    .input(UpdateHazardReportStatusSchema)
    .mutation(async ({ input, ctx }) => {
      const now = new Date()
      const isResolution = input.status === 'resolved' || input.status === 'dismissed'

      await ctx.db
        .update(hazardReports)
        .set({
          status: input.status,
          reviewedBy: ctx.user.id,
          reviewNote: input.reviewNote ?? null,
          updatedAt: now,
          resolvedAt: isResolution ? now : null,
        })
        .where(eq(hazardReports.id, input.id))

      await ctx.db.insert(auditLogs).values({
        userId: ctx.user.id,
        action: `report.${input.status}`,
        details: { reportId: input.id, reviewNote: input.reviewNote ?? null },
      })

      return { success: true }
    }),

  stats: protectedProcedure.query(async ({ ctx }) => {
    const conditions = []

    if (ctx.user.role !== 'admin') {
      const subscription = await ctx.db.query.alertSubscriptions.findFirst({
        where: eq(alertSubscriptions.userId, ctx.user.id),
      })
      const subscribedZones = subscription?.zones ?? []
      if (subscribedZones.length > 0) {
        conditions.push(inArray(hazardReports.zone, subscribedZones))
      }
    }

    const where = conditions.length > 0 ? and(...conditions) : undefined

    const rows = await ctx.db
      .select({ status: hazardReports.status, count: count() })
      .from(hazardReports)
      .where(where)
      .groupBy(hazardReports.status)

    const byStatus: Record<string, number> = {}
    for (const row of rows) {
      byStatus[row.status] = row.count
    }

    const pending = byStatus['pending'] ?? 0
    const reviewing = byStatus['reviewing'] ?? 0
    const resolved = byStatus['resolved'] ?? 0
    const dismissed = byStatus['dismissed'] ?? 0

    return { pending, reviewing, resolved, dismissed, total: pending + reviewing + resolved + dismissed }
  }),
})

File: crates/scemas-desktop/src/queries/reports.rs
use chrono::{DateTime, Utc};
use serde::Serialize;
use sqlx::PgPool;
use uuid::Uuid;

#[derive(Debug, Serialize, sqlx::FromRow)]
#[serde(rename_all = "camelCase")]
pub struct HazardReport {
    pub id: Uuid,
    pub zone: String,
    pub category: String,
    pub description: String,
    pub status: String,
    pub contact_email: Option<String>,
    pub reported_by: Option<Uuid>,
    pub reviewed_by: Option<Uuid>,
    pub review_note: Option<String>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
    pub resolved_at: Option<DateTime<Utc>>,
}

pub async fn list(
    pool: &PgPool,
    status: Option<&str>,
    limit: i64,
) -> Result<Vec<HazardReport>, sqlx::Error> {
    sqlx::query_as::<_, HazardReport>(
        "SELECT * FROM hazard_reports
         WHERE ($1::text IS NULL OR status = $1)
         ORDER BY created_at DESC
         LIMIT $2",
    )
    .bind(status)
    .bind(limit)
    .fetch_all(pool)
    .await
}

pub async fn submit(
    pool: &PgPool,
    zone: &str,
    category: &str,
    description: &str,
    contact_email: Option<&str>,
    reported_by: Option<Uuid>,
) -> Result<HazardReport, sqlx::Error> {
    sqlx::query_as::<_, HazardReport>(
        "INSERT INTO hazard_reports (zone, category, description, contact_email, reported_by)
         VALUES ($1, $2, $3, $4, $5)
         RETURNING *",
    )
    .bind(zone)
    .bind(category)
    .bind(description)
    .bind(contact_email)
    .bind(reported_by)
    .fetch_one(pool)
    .await
}

pub async fn update_status(
    pool: &PgPool,
    id: Uuid,
    status: &str,
    review_note: Option<&str>,
    reviewed_by: Option<Uuid>,
) -> Result<(), sqlx::Error> {
    ::sqlx::query(
        "UPDATE hazard_reports
         SET status = $2,
             review_note = $3,
             reviewed_by = $4,
             updated_at = NOW(),
             resolved_at = CASE WHEN $2 = 'resolved' THEN NOW() ELSE resolved_at END
         WHERE id = $1",
    )
    .bind(id)
    .bind(status)
    .bind(review_note)
    .bind(reviewed_by)
    .execute(pool)
    .await?;
    Ok(())
}
